devices/flash-zynq7000/qspi.c

```diff
@@ -0,0 +1,329 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system loader
+ *
+ * Quad-SPI Controller driver
+ *
+ * Copyright 2020 Phoenix Systems
+ * Author: Hubert Buczynski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include "qspi.h"
+#include <lib/errno.h>
+
+
+enum { cr = 0, sr, ier, idr, imr, er, dr, txd00, rxd, sicr, txth, rxth, gpio,
+	   lpbk = 0xe, txd01 = 0x20, txd10, txd11,
+	   lqspi_cr = 0x28, lqspi_sr, modid = 0x3f };
+
+
+struct {
+	volatile u32 *base;
+} qspi_common;
+
+
+void qspi_stop(void)
+{
+	u32 data;
+
+	/* Clean up RX Fifo */
+	while ((*(qspi_common.base + sr) & (1 << 4)))
+		data = *(qspi_common.base + rxd);
+
+	*(qspi_common.base + cr) |= (1 << 10);
+	*(qspi_common.base + er) = 0;
+}
+
+
+void qspi_start(void)
+{
+	*(qspi_common.base + rxth) = 0x1;
+	*(qspi_common.base + cr) &= ~(1 << 10);
+	*(qspi_common.base + er) = 0x1;
+}
+
+
+static unsigned int qspi_rxData(u8 *rxBuff, size_t size)
+{
+	const u32 data = *(qspi_common.base + rxd);
+
+	if (rxBuff != NULL) {
+		switch (size) {
+			case 0:
+				return 0;
+
+			case 1:
+				rxBuff[0] = (data >> 24) & 0xff;
+				return 1;
+
+			case 2:
+				rxBuff[0] = (data >> 16) & 0xff;
+				rxBuff[1] = (data >> 24) & 0xff;
+				return 2;
+
+			case 3:
+				rxBuff[0] = (data >> 8) & 0xff;
+				rxBuff[1] = (data >> 16) & 0xff;
+				rxBuff[2] = (data >> 24) & 0xff;
+				return 3;
+
+			case 4:
+			default:
+				rxBuff[0] = data & 0xff;
+				rxBuff[1] = (data >> 8) & 0xff;
+				rxBuff[2] = (data >> 16) & 0xff;
+				rxBuff[3] = (data >> 24) & 0xff;
+				return 4;
+		}
+	}
+
+	return (size < 4) ? size : 4;
+}
+
+
+static unsigned int qspi_txData(const u8 *txBuff, size_t size)
+{
+	const u32 dummy = 0;
+	const u8 *buff = (txBuff == NULL) ? (const u8 *)&dummy : txBuff;
+
+	switch (size) {
+		case 0:
+			return 0;
+
+		case 1:
+			*(qspi_common.base + txd01) = buff[0];
+			return 1;
+
+		case 2:
+			*(qspi_common.base + txd10) = buff[0] | (buff[0] >> 8);
+			return 2;
+
+		case 3:
+			*(qspi_common.base + txd11) = buff[0] | (buff[0] >> 8) | (buff[0] >> 16);
+			return 3;
```

Comment:
Is this code correct?
`buf[0]` is of type `u8` and shift it to the right by 8 or 16 gives `0`.

Reply:
Nope this code is not correct. Thank You for noting it!!

It's just an ugly copy-paste from the other function (from the code developing), which cause that I have left it as it is. While I have tested the code I usually have sent data to program with size `0x100` therefore it have't gone to these cases.
When I have sent status commands, it has went to these cases but sent the dummy data so it looked ok. 
To sump up,  it was critical error, so thanks again for pointing it out. 

Fixed.